eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
-     
-     (list (+ (- x-exp out-exp) slack 1))]
+     (list (- (maxlog x) (minlog z)))]
+
+    [(ival-cos)
+     ; maxlog(x) - minlog(z) + min(maxlog(x), 0)
+     (define x (first srcs))
+     (list (+ (- (maxlog x) (minlog z)) (min (maxlog x) 0)))]
+
+    [(ival-sinh)
+     ; maxlog(x) + logspan(z) - min(minlog(x), 0)
+     (define x (first srcs))
+     (list (- (+ (maxlog x) (logspan z)) (min (minlog x) 0)))]
+
+    [(ival-cosh)
+     ; maxlog(x) + logspan(z) + min(maxlog(x), 0)
+     (define x (first srcs))
+     (list (+ (maxlog x) (logspan z) (min (maxlog x) 0)))]
     
     [(ival-log ival-log2 ival-log10)
-     ; log[Гlog]   = log[1/logx] = -log[log(x)]
-     ; log[Гlog2]  = log[1/(log2(x) * ln(2))] <= -log[log2(x)] + 1    < main formula
-     ; log[Гlog10] = log[1/(log10(x) * ln(10))] <= -log[log10(x)] - 1
-     ; log[Гlog10] < log[Гlog] < log[Гlog2]
-     
-     (define outlo (ival-lo output))
-     (define outhi (ival-hi output))
-     (define out-exp (- (ival-min-log2-approx output))) ; -log[log10(x)]
-     
-     (define slack (if (equal? (mpfr-sign outlo) (mpfr-sign outhi))
-                       0
-                       (get-slack)))                  ; output crosses 0 - uncertainty
-     
-     (list (+ out-exp 1 slack))]
-    
+     ; log:   logspan(x) - minlog(z)
+     ; log2:  logspan(x) - minlog(z) + 1
+     ; log10: logspan(x) - minlog(z) - 1
+     (define x (first srcs))
+     (list (+ (- (logspan x) (minlog z)) 1))]
+
     [(ival-asin ival-acos)
-     ; log[Гasin] = log[x] - log[1-x^2]/2 - log[asin(x)] + 1
-     ; log[Гacos] = log[x] - log[1-x^2]/2 - log[acos(x)] + 1
-     ;                       ^^^^^^^^^^^^
-     ;                       condition of uncertainty
+     ; asin: maxlog(x) - log[1-x^2]/2 - minlog(z)
+     ; acos: maxlog(x) - log[1-x^2]/2 - minlog(z)
+     ;                   ^^^^^^^^^^^^
+     ;              condition of uncertainty
      (define x (first srcs))
-     (define x-exp (ival-max-log2-approx x))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (define slack (if (>= x-exp 1)                   ; Condition of uncertainty when argument > sqrt(3)/2
+     (define slack (if (>= (maxlog x) 1)              ; Condition of uncertainty when argument > sqrt(3)/2
                        (get-slack)                    ; assumes that log[1-x^2]/2 is equal to slack
                        0))
-     
-     (list (+ (- x-exp out-exp) 1 slack))]
+     (list (+ (- (maxlog x) (minlog z)) slack))]
     
     [(ival-atan)
-     ; log[Гatan] = log[x] - log[x^2+1] - log[atan(x)] <= -|log[x]| - log[atan(x)] + 1 <= 0
+     ; logspan(x) - |minlog(x)| - minlog(z)
      (define x (first srcs))
-     
-     (define x-exp-abs (max (abs (log2-approx (ival-hi x)))
-                            (abs (log2-approx (ival-lo x)))))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (list (+ (- (- x-exp-abs) out-exp) 1))]
-    
+     (list (- (logspan x) (abs (minlog x)) (minlog z)))]
```

Comment:
I'm suspicious of `(abs (minlog x))`. Shouldn't it be something like `min(abs(minlog(x), maxlog(x)))`? Not sure.

Reply:
Yes, I think so too